### PR TITLE
[Dynamic Shapes] Rebind unbacked symbols to derived expressions

### LINKED ISSUE #183677: [HOP] Alias/mutation fake traces should ignore discarded unbacked symbols
## Summary

FlexAttention can fail during HOP metadata/functionalization when it is used with dynamic block-mask inputs.

The relevant path is:

`torch.compile(flex_attention)` -> `flex_attention_hop` -> FlexAttention functionalization -> score_mod alias/mutation query -> nested fake trace.

That nested trace is only a diagnostic query: it answers whether `score_mod` may mutate captured inputs. The graph is discarded. With dynamic block-mask metadata, fake propagation can leave fresh unbacked symbols pending in the surrounding `ShapeEnv`. Those symbols are not semantic outputs of the score-mod alias/mutation query, but `compute_unbacked_bindings` can still require them to appear in the returned value and raise `PendingUnbackedSymbolNotFound`.

## Reproducer

```python
import torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention


if not torch.cuda.is_available():
    raise RuntimeError("This repro needs CUDA because FlexAttention backward is CUDA-only.")


device = "cuda"
dtype = torch.float16


def score_mod(score, batch, head, q_idx, kv_idx):
    return score


compiled_create_block_mask = torch.compile(
    create_block_mask,
    dynamic=True,
    fullgraph=True,
)


def create_dynamic_block_mask(q_batch, kv_batch):
    q_len = q_batch.size(0)
    kv_len = kv_batch.size(0)

    def mask_mod(batch, head, q_idx, kv_idx):
        q_group = q_batch[q_idx]
        kv_group = kv_batch[kv_idx]
        return (q_group == kv_group) & (q_group != -1) & (kv_group != -1)

    return compiled_create_block_mask(
        mask_mod,
        B=None,
        H=None,
        Q_LEN=q_len,
        KV_LEN=kv_len,
        device=device,
        BLOCK_SIZE=128,
    )


groups = torch.zeros(128, dtype=torch.int64, device=device)
block_mask = create_dynamic_block_mask(groups, groups)

q = torch.randn(1, 1, 128, 64, device=device, dtype=dtype, requires_grad=True)
k = torch.randn(1, 1, 128, 64, device=device, dtype=dtype, requires_grad=True)
v = torch.randn(1, 1, 128, 64, device=device, dtype=dtype, requires_grad=True)

compiled_flex_attention = torch.compile(
    flex_attention,
    fullgraph=True,
    dynamic=True,
    backend="aot_eager",
)

out = compiled_flex_attention(q, k, v, score_mod=score_mod, block_mask=block_mask)
out.sum().backward()
torch.cuda.synchronize()
print("ok", out.shape)
```

## Expected behavior

The compiled FlexAttention forward and backward should complete. The score-mod alias/mutation query should be allowed to discard fresh unbacked symbols created only for HOP metadata/fake propagation.

## Actual behavior

The FlexAttention HOP metadata path can fail with:

```text
torch.fx.experimental.symbolic_shapes.PendingUnbackedSymbolNotFound: Pending unbacked symbols {...} not in returned outputs ...
```

The pending symbols are not part of the user-visible output. They are created during the internal fake/metadata trace used by the HOP alias/mutation check.

## Suggested fix

The internal HOP trace should run under the existing `ignore_fresh_unbacked_symbols` contract, and `compute_unbacked_bindings` should honor that contract by treating pending symbols as ignorable while that TLS is active.

The HOP alias/mutation helper should detect an existing fake mode from inputs, and run its diagnostic `make_fx` trace under the ignore-fresh-unbacked context, including when `make_fx` creates its own fake mode. This keeps diagnostic fake traces from imposing output-binding requirements on symbols that are intentionally discarded with the diagnostic graph.

### PR DESCRIPTION
**Pull request by sanketpurandare**
`rebind_unbacked` already records equivalences when a retraced binding site maps
an unbacked symbol to another symbol or to a constant. The same invariant applies
when the new value is a derived symbolic expression such as `(u1 + 1) // 2`: the
old symbol still has a concrete binding relationship and should be eliminated in
favor of that expression.

The previous assertion assumed any non-symbol replacement with free symbols was
invalid. That is too strong for legitimate derived unbacked shapes and makes the
binding logic reject a value it can represent with the existing ShapeEnv
replacement mechanism. Record the replacement with `_eliminate_unbacked`, which
is the existing path for replacing an unbacked symbol by a non-symbol expression.

This intentionally does not restore the reverted broad HOP fake-trace
suppression. The FlexAttention/HOP reproducer for #183677 now passes on current
main without that suppression, and the broad suppression was the source of the
internal cond, AOTInductor, Executorch, and FlexAttention regressions.

This change was authored with assistance from an AI assistant.

Test Plan:

```bash
python test/test_dynamic_shapes.py TestUnbacked.test_rebind_unbacked_to_symbolic_expression -v
```

```bash
python test/dynamo/test_misc.py MiscTests.test_cond_runtime_assert_generation -v
```

```bash
python test/functorch/test_control_flow.py TestControlFlowTraced.test_cond_functionalized_nested -v
```

```bash
python test/inductor/test_aot_inductor.py -k cond_non_tensor_predicates_dynamic_True_cpu -v
```

```bash
python - <<'PY'
import torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention

if not torch.cuda.is_available():
    raise RuntimeError('CUDA unavailable')

device = 'cuda'
dtype = torch.float16

def score_mod(score, batch, head, q_idx, kv_idx):
    return score

compiled_create_block_mask = torch.compile(create_block_mask, dynamic=True, fullgraph=True)

def create_dynamic_block_mask(q_batch, kv_batch):
    q_len = q_batch.size(0)
    kv_len = kv_batch.size(0)
    def mask_mod(batch, head, q_idx, kv_idx):
        q_group = q_batch[q_idx]
        kv_group = kv_batch[kv_idx]
        return (q_group == kv_group) & (q_group != -1) & (kv_group != -1)
    return compiled_create_block_mask(mask_mod, B=None, H=None, Q_LEN=q_len, KV_LEN=kv_len, device=device, BLOCK_SIZE=128)

groups = torch.zeros(128, dtype=torch.int64, device=device)
block_mask = create_dynamic_block_mask(groups, groups)
q = torch.randn(1, 1, 128, 64, device=device, dtype=dtype, requires_grad=True)
k = torch.randn(1, 1, 128, 64, device=device, dtype=dtype, requires_grad=True)
v = torch.randn(1, 1, 128, 64, device=device, dtype=dtype, requires_grad=True)
compiled_flex_attention = torch.compile(flex_attention, fullgraph=True, dynamic=True, backend='aot_eager')
out = compiled_flex_attention(q, k, v, score_mod=score_mod, block_mask=block_mask)
out.sum().backward()
torch.cuda.synchronize()
print('ok', out.shape)
PY
```

```bash
lintrunner -a
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1443,6 +1443,11 @@ def compute_unbacked_bindings(
 
     pending = set(fresh_sym)
     ignorable = set(ign_sym)
+    if shape_env._ignore_fresh_unbacked_symbols_tls():
```
Comment (laithsakka) [inline torch/fx/experimental/symbolic_shapes.py]: This does not seem right? if those are allocated before the we set the ignore then we should bound them otherwise they shall not be added to pending

Comment (laithsakka) [reply torch/fx/experimental/symbolic_shapes.py]: are we missing a call to compute_unbacked_bindings? or do we want stack them /add them back on a context manager. 
I am not sure about your issue exactly what happen can you explain whats going on.

File: torch/_higher_order_ops/utils.py [hunk 2]
```diff
@@ -348,12 +358,24 @@ def _maybe_fake_tracing(fn, inputs: list[Any], pre_dispatch):
...
-            pre_dispatch=pre_dispatch,
-            _error_on_data_dependent_ops=False,
-        )(*inputs)
+        from torch.fx.experimental.symbolic_shapes import ShapeEnv
+
+        shape_env = fake_mode_det.shape_env if fake_mode_det is not None else None
+        ignore_fresh_unbacked = (
+            shape_env.ignore_fresh_unbacked_symbols()
+            if shape_env is not None
+            else ShapeEnv().ignore_fresh_unbacked_symbols()
```
Comment (laithsakka) [inline torch/_higher_order_ops/utils.py]: ShapeEnv().ignore_fresh_unbacked_symbols() is odd — constructing a ShapeEnv purely to flip a thread-global flag is misleading and wasteful.

maybe introduce  ignore_fresh_unbacked_symbols_tls() context manager , even better if you remove all current usages of shape_env.ignore_fresh_unbacked_symbols()
rmoving the indirection makes it obvious that the flag is thread-global state, not per-ShapeEnv state.

**Review by laithsakka — CHANGES_REQUESTED** [review]

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (sanketpurandare) [reply torch/fx/experimental/symbolic_shapes.py]: Updated this so the ignore scope only suppresses symbols created inside the temporary fake/metadata trace. New unbacked symbols created while the TLS is active are not added to `pending_fresh_unbacked_symbols`; `compute_unbacked_bindings` no longer treats all pending symbols as ignorable just because the TLS flag is active. I also added a focused test that creates one pending symbol before the ignore scope and one inside it, then proves the pre-existing pending symbol still has to bind to the returned value.

File: torch/_higher_order_ops/utils.py (hunk 2 above)
Comment (sanketpurandare) [reply torch/_higher_order_ops/utils.py]: Removed the dummy ShapeEnv construction and made the thread-local state explicit. `_maybe_fake_tracing` now uses `ignore_fresh_unbacked_symbols_tls()` directly, so the code no longer suggests that this is per-ShapeEnv mutable state.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -3767,6 +3767,32 @@ def relation_with_digit(expr: str, op: str, digit: int) -> None:
 TLS = threading.local()
 
 
+def _ignore_fresh_unbacked_symbols_tls() -> bool:
```
Comment (laithsakka) [inline torch/fx/experimental/symbolic_shapes.py:3777]: nit: just inline it have one user? its confusing

**Review by laithsakka — APPROVED** [review]
thank you

**Comment by wdvr** [comment]
@pytorchmergebot revert -m "failing some executorch / other tests on the HOP / control-flow / FlexAttention surface I believe" -c ghfirst

claude analysis:

36 × pytorch/tritonbench/.../test_operator_accuracy[_beta[_triton_beta]] - test_operator_accuracy_*_flex_attention_accuracy_* (noop, causal, head_bias, sliding_window, document_mask, prefix_lm, softcap) — accuracy/listing failures across multiple error_clusters 
-> Related (the diff explicitly names FlexAttention's score_mod alias check as the motivating fix path).

43 × caffe2/test/inductor:test_aot_inductor - test_cond_* / test_while_loop_* / test_custom_op_in_subgraph_* cpu+cuda — clusters 1017592810737058, 1161238703736894, 1267440348546219 -> testx Good Quality with 4 retries each;
-> Related (HOP cond/while_loop AOT inductor codegen path directly exercises the modified compute_unbacked_bindings and HOP alias/mutation helper).

9 × executorch/exir/tests:passes + executorch/exir/backend/test:test_backends_* + executorch/exir/tests:memory_planning / test_constant_prop_pass_for_control_flow, test_debug_handle_generator_pass_with_control_flow, test_spec_prop_pass_cond, test_alloc_node_spec, test_partition_with_control_flow, test_partition_with_map, test_multi_map, test, test_spec_prop_pass_scan
-> Related (executorch control-flow / map passes exercise the same HOP infrastructure).

7 × executorch/exir/emit/test:emit + executorch/extension/llm/custom_ops:test_update_cross_attn_cache + executorch/test/end2end:end2end_aten — test_emit_cond, test_emit_map, test_load_emit_map, test_emit_scan_add_mul, test_run_emit_scan_cumsum, test_run_emit_scan_dynamic_shape, test_update_cross_attn_cache_in_cond, test_ft_map_dynshape
